File: masonry/src/engine/tmux/wave.js
'use strict';
// engine/tmux/wave.js — Batch spawn and collect for agent waves.
//
// Port of bl/tmux/wave.py to Node.js.

const { execFileSync } = require('child_process');
const { spawnAgent } = require('./core');
const { waitForAgent } = require('./core');
const { inTmux } = require('./helpers');

function spawnWave(agents, { maxConcurrency = null, spawner = null } = {}) {
  const batch = maxConcurrency ? agents.slice(0, maxConcurrency) : agents;
  const _spawn = spawner || ((name, prompt, opts) => spawnAgent(name, prompt, opts));

  const spawns = [];
  for (const spec of batch) {
    const { agentName, prompt, ...rest } = spec;
    spawns.push(_spawn(agentName, prompt, rest));
  }

  if (inTmux() && spawns.length) {
    try {
      execFileSync('tmux', ['select-layout', 'tiled'], {
        timeout: 5000,
        stdio: ['pipe', 'pipe', 'pipe'],
      });
    } catch {
      // ignore layout failures
    }
  }

  return spawns;
}

function collectWave(spawns, { timeout = 600, waiter = null } = {}) {
  const _wait = waiter || ((spawn, opts) => waitForAgent(spawn, opts));
  return spawns.map(s => _wait(s, { timeout }));
}

module.exports = {
  spawnWave,
  collectWave,
};

File: bl/tmux/wave.py
"""bl/tmux/wave.py — Batch spawn and collect for agent waves."""

from __future__ import annotations

import subprocess
from bl.tmux.core import AgentResult, SpawnResult, spawn_agent, wait_for_agent
from bl.tmux.helpers import in_tmux


def spawn_wave(
    agents: list[dict[str, object]],
    max_concurrency: int | None = None,
) -> list[SpawnResult]:
    """Spawn a batch of agents. Returns list of SpawnResults for collection.

    Each dict in agents must have 'agent_name' and 'prompt'; other keys are
    forwarded as kwargs to spawn_agent.
    """
    batch = agents[:max_concurrency] if max_concurrency else agents

    spawns: list[SpawnResult] = []
    for spec in batch:
        spawns.append(spawn_agent(**spec))  # pyright: ignore[reportArgumentType]

    if in_tmux() and spawns:
        try:
            _ = subprocess.run(
                ["tmux", "select-layout", "tiled"],
                capture_output=True,
                timeout=5,
            )
        except (subprocess.TimeoutExpired, subprocess.CalledProcessError):
            pass

    return spawns


def collect_wave(
    spawns: list[SpawnResult],
    *,
    timeout: int = 600,
) -> list[AgentResult]:
    """Wait for all spawns and return their results."""
    return [wait_for_agent(s, timeout=timeout) for s in spawns]
